### ISSUE TITLE
Tone detection

### ISSUE THREAD

**Issue by YoannQueret** [savonet/liquidsoap#1796]
Hi !

I think there is no functionality to detect a tone or DTMF (dual-tone multi-frequency) in liquidsoap ?

Is this a feature that could be considered?

- Detection of a tone (or dual-tone)
- Detection of a sequence of tone (or dual-tone)
- Parameters : Minimum length of a tone
- Parameters : Maximum delays between each tone in the case of a sequence of tone
- Call an handler for each detection

For example :
- on detection of the 1750Hz tone -> start an output
- on detection of a DTMF * 56 -> play a specific audio file
- etc ...

What do you think ?

(This requirement is for a ham-radio project)

Very good week
Yoann

**Comment by smimram** [comment]
Yes, it should not be too hard if you have use for it!

**Comment by YoannQueret** [comment]
Hi Samuel,
Many thank for this new function.
As i wrote, i need a tone detection on an audio  source, not generator (but can be also very usefull)
I think it's a little more complicated, so do you think it's possible?
Yoann

**Comment by smimram** [comment]
Yes, I know :) I started reading about DTMF and thought I'd start with a generator, which will be needed to test the detector anyway. But I need some more time to understand the Goertzel algorithm, which seems to be the standard way of performing this task...

**Comment by smimram** [comment]
I have just merged the implementation of `dtmf.detect` into main. It works on my test, but the parameters might need some fine-tuning depending on the situations where you want to use this. Could you test it and keep me updated? A simple example is
```
def f(k)
  print("Detected key #{k}")
end

s = dtmf(duration=0.1, "1234567890*#ABCD")
s = amplify(0.5, s)
s = add([s, amplify(0.5, noise())])
s = dtmf.detect(s, f)

output(s)
```

**Comment by pteich** [comment]
@smimram I just noticed that you implemented the Goertzel algorithm in Liquidsoap. Great! Do you think this could also be used to detect a single specific frequency beside of DTMF? In my case a 18Hz sinus tone used to signal events in radio streams.

**Comment by smimram** [comment]
@pteich yes it should be not too difficult to adapt the code!

**Comment by toots** [comment]
Would be nice it you could take a minute to look at https://github.com/savonet/liquidsoap/issues/1904 if you're at it again :-)

**Comment by smimram** [comment]
Ok so both issues should hopefull be addressed in #1945.

**Comment by pteich** [comment]
Amazing work @smimram - 1 hour from comment to PR :) I'll check this out, it makes our workflow a lot easier. Merci!
